### ISSUE TITLE
Migrate to Next.js

### ISSUE THREAD

**Issue by ezhlobo** [datarockets/dreact#25]
There are two key advantages for us that Next.js has over CRA:
- Built in server side rendering
- Better optimization

Due to them I think it makes sense to run an experiment and see if we can build our framework on top of Next.js that covers our needs.

---

## Progress

*Checklist will be here*

**Comment by ezhlobo** [comment]
Next.js relies on react and react-dom packages in Node scripts. It's possible to mock them and point to necessary files, however it seems like they are run in `jest-worker` and it can be quite tricky to achieve and override the references.

**Comment by nixololo** [comment]
@ezhlobo hey, do you need help with it?

**Comment by ezhlobo** [comment]
@nixololo my main issue is that I don't know how to hide `next` dependency behind react.

Ideally I don't want to expose what is used under the hood and provide own framework, so it's important.

**Comment by ezhlobo** [comment]
I attached working branch. I successfully ran nextjs behind `dreact app`. However I stuck on a task when I need to sync server side router and local router.

- [x] Local development
- [x] Nodejs server that does SSR
- [x] Static app (when all files are plain html files)
- [ ] Dynamic app (when there's one html file that handles all routes)

I want to get a router that can be transformed into something that gives the possibility to implement steps mentioned above.

Also it seems like we need to invent new commands. I'm thinking about something like this:

- **Local development**
	`dreact app dev`
- **Nodejs server that does SSR**
	`dreact app ssr-build`
	`dreact app ssr`
- **Static app**
	`dreact app static-build`
- **Dynamic app**
	`dreact app dynamic-build`

PS. Even though my main focus is to implement local development and dynamic app, I need to be careful and pay attention on all the constraints we have from SSR and Static app.

**Comment by viscoze** [comment]
@ezhlobo when you're talking about "Dynamic app" do you mean implementing of client side routing? 

Like we receive an html with one script tag, and then receive a bundle of the whole app. Then we use routing to navigate to other routes directly in a browser

**Comment by ezhlobo** [comment]
@viscoze yes, "Dynamic apps" are those that understand what to output in runtime, not on server or via pre-building all pages.
